stablehlo/reference/Ops.cpp

```diff
@@ -133,6 +322,182 @@ Tensor evalConvertOp(const Tensor &operand, TensorType resultType) {
+  auto lhsWindowDilations = concatAndPermute(
+      1L, llvm::to_vector(rhsDilation), 1L, inputBatchDimension,
+      inputSpatialDimensions, inputFeatureDimension);
+
+  auto outputSpacialIndexIt = IndexSpaceIterator(
+      Sizes(extractElements(result.getShape(), outputSpatialDimensions)),
+      Index(outputSpatialDimensions.size()));
+  auto outputSpacialIndexItEnd = IndexSpaceIterator(
+      Sizes(extractElements(result.getShape(), outputSpatialDimensions)),
+      std::nullopt);
+  for (; outputSpacialIndexIt != outputSpacialIndexItEnd;
+       ++outputSpacialIndexIt) {
+    SmallVector<int64_t> lhsPaddingLow;
+    for (auto paddingPair : lhsPadding)
+      lhsPaddingLow.push_back(paddingPair.first);
+    auto paddedLhs =
+        evalPadOp(lhs, getZeroScalarTensor(result.getElementType()),
+                  Sizes(lhsPaddingLow), Sizes(lhsBaseDilations),
+                  inferPadOpType(lhsPadding, lhs.getType(),
+                                 result.getElementType(), lhsBaseDilations));
+
+    SmallVector<int64_t> lhsWindowStart;
+    for (auto [i, offset] : llvm::enumerate(
+             concatAndPermute(0L, llvm::to_vector(*outputSpacialIndexIt), 0L,
+                              inputBatchDimension, inputSpatialDimensions,
+                              inputFeatureDimension)))
+      lhsWindowStart.push_back(lhsWindowStrides[i] * offset);
+    SmallVector<int64_t> limitIndices;
+    for (size_t i = 0; i < lhsWindowStart.size(); ++i)
+      limitIndices.push_back(std::min(
+          lhsWindowStart[i] + lhsWindowDimensions[i] * lhsWindowDilations[i],
+          paddedLhs.getShape()[i]));
+    auto lhsWindow =
+        evalSliceOp(paddedLhs, Sizes(lhsWindowStart), Sizes(lhsWindowDilations),
+                    inferSliceOpType(paddedLhs.getType(), lhsWindowStart,
+                                     limitIndices, lhsWindowDilations));
+
+    SmallVector<int64_t> reverseDims;
+    for (auto [i, isReverse] : llvm::enumerate(windowReversal))
+      if (isReverse) reverseDims.push_back(inputSpatialDimensions[i]);
+    auto reversedLhsWindow =
+        evalReverseOp(lhsWindow, Axes(reverseDims), lhsWindow.getType());
+
+    auto lhsContractingDimensions = llvm::to_vector(inputSpatialDimensions);
+    lhsContractingDimensions.push_back(
+        static_cast<int64_t>(inputFeatureDimension));
+    auto rhsContractingDimensions = llvm::to_vector(kernelSpatialDimensions);
+    rhsContractingDimensions.push_back(
+        static_cast<int64_t>(kernelInputFeatureDimension));
+    auto dotProduct = evalDotGeneralOp(
+        reversedLhsWindow, rhs, /*lhsBatchingDimensions=*/{},
+        /*rhsBatchingDimensions=*/{}, Axes(lhsContractingDimensions),
+        Axes(rhsContractingDimensions),
+        inferDotGeneralOpType(reversedLhsWindow.getType(), rhs.getType(),
+                              lhsContractingDimensions,
+                              rhsContractingDimensions));
+
+    auto resultShape = [&](auto resultOffsetIt) {
```

Comment:
We can skip this  `resultShape` lambda and re-use  `concatAndPermute` instead as that would rhyme better with spec
```
* `result_shape(n1, hw, c1) = permute([n1] + hw + [c1], [output_batch_dimension] + output_spatial_dimensions + [output_feature_dimension])`.
```

Using something like

```
    Sizes resultNonSpatialDims;
    for (auto i = 0; i < result.getRank(); ++i)
      if (llvm::find(outputSpatialDimensions, i) ==
          outputSpatialDimensions.end())
        resultNonSpatialDims.push_back(result.getShape()[i]);

    auto resultPermutation = output_batch_dimension + output_spatial_dimensions + output_feature_dimension;
    auto resultNonSpatialIt =
        IndexSpaceIterator(resultNonSpatialDims, Index(resultNonSpatialDims.size()));
    for (auto dotProductIt = dotProduct.index_begin();
         dotProductIt != dotProduct.index_end();
         ++dotProductIt, ++resultNonSpatialIt) {
      auto resultIndex = concatAndPermute((*resultNonSpatialIt)[0], *outputSpacialIndexIt, (*resultNonSpatialIt)[1], resultPermutation);
      result.set(Index(resultIndex), dotProduct.get(*dotProductIt));
    }

```

Reply:
> auto resultPermutation = output_batch_dimension + output_spatial_dimensions + output_feature_dimension;

As noted in another comment, we can't quite use the syntax sugar above, but I've refactored the helper to be more aligned with the spec.